### ISSUE TITLE
Questions about the write length of lws_write

### ISSUE THREAD

**Issue by lestLyy** [warmcat/libwebsockets#3118]
You can often see the following writing.
```c
char buf[LWS_PRE + 4096], *start = &buf[LWS_PRE], *p = start, *end = &buf[sizeof(buf) - 1];
p += lws_snprintf(p, lws_ptr_diff_size_t(end, p), pucContent + uiSendIndex);
if (lws_write(wsi, (uint8_t *)start, lws_ptr_diff_size_t(p, start), (enum lws_write_protocol)LWS_WRITE_HTTP) != lws_ptr_diff(p, start))
    ...
```
I split this big data into 4096 bytes to send, but lws_ptr_diff(p, start) is not the real len that i send, lws_ptr_diff(p, start) is One larger than the strlen(start). Does lws_ptr_diff (p, start) add \ 0 at the end

**Comment by lws-team** [comment]
No, lws_ptr_diff(a, b) is just a typesafe helper that does effectively (((uint8_t *)b) - ((uint8_t *)a)).  First it's setting the limit for safely doing the lws_snprintf() write into the buffer, and for the second usage it's figuring out how much of the buffer we used.

The usage of lws_snprintf() is insecure, since passing the string to "print" into the buffer in the format parameter means that if it contains format modifiers like %s, it will include things from the stack.  You should change it to ... "%s", pucContent + uiSendIndex);

Otherwise I don't think it should go wrong, you should log exactly what the offset stuff is doing each time around.
